File: zig/basics/04_data_structures.zig
const std = @import("std");
// TOPIC: Data Structures | zig run 04_data_structures.zig

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // TODO 1: ArrayList — dynamic array
    //   var list = std.ArrayList(i32).init(allocator);
    //   defer list.deinit();
    //   try list.append(1); try list.append(2); try list.append(3);
    //   list.items[0]         // 1
    //   list.items.len        // 3
    //   try list.insert(1, 99);  // insert at index 1
    //   _ = list.orderedRemove(0);  // remove at index 0
    //   list.clearRetainingCapacity();  // clear without dealloc

    // TODO 2: StringHashMap — string keys
    //   var map = std.StringHashMap(i32).init(allocator);
    //   defer map.deinit();
    //   try map.put("alice", 30);
    //   try map.put("bob", 25);
    //   const val = map.get("alice");  // ?i32 — optional
    //   var iter = map.iterator();
    //   while (iter.next()) |entry| {
    //       std.debug.print("{s}: {d}\n", .{entry.key_ptr.*, entry.value_ptr.*});
    //   }

    // TODO 3: AutoHashMap — infer key type
    //   var intMap = std.AutoHashMap(i32, []const u8).init(allocator);
    //   defer intMap.deinit();
    //   try intMap.put(1, "one"); try intMap.put(2, "two");

    // TODO 4: Linked list with allocator
    //   const Node = struct {
    //       value: i32,
    //       next: ?*Node = null,
    //       fn create(allocator: std.mem.Allocator, value: i32) !*Node {
    //           const node = try allocator.create(Node);
    //           node.* = .{ .value = value };
    //           return node;
    //       }
    //   };
    //   var head = try Node.create(allocator, 1);
    //   defer allocator.destroy(head);
    //   head.next = try Node.create(allocator, 2);
    //   defer allocator.destroy(head.next.?);

    // TODO 5: Stack using ArrayList
    //   var stack = std.ArrayList(i32).init(allocator);
    //   defer stack.deinit();
    //   try stack.append(1);    // push
    //   _ = stack.pop();        // pop (returns ?T — null if empty)
    //   stack.getLast()         // peek

    // TODO 6: Queue using TailQueue
    //   var queue = std.TailQueue(i32){};
    //   var node1 = std.TailQueue(i32).Node{ .data = 1 };
    //   queue.append(&node1);
    //   queue.popFirst()  // dequeue

    // TODO 7: MultiArrayList — struct-of-arrays for cache efficiency
    //   const Point = struct { x: f32, y: f32 };
    //   var points = std.MultiArrayList(Point){};
    //   defer points.deinit(allocator);
    //   try points.append(allocator, .{ .x = 1.0, .y = 2.0 });
    //   const xs = points.items(.x);  // []f32 — all x values contiguous in memory

    // TODO 8: BoundedArray — fixed-capacity stack-allocated array
    //   var buf = std.BoundedArray(i32, 10){};
    //   try buf.append(1); try buf.append(2);
    //   buf.slice()  // []i32

    std.debug.print("TODO: implement data structure exercises\n", .{});
    _ = allocator;
}

File: ocaml/basics/04_data_structures.ml
(* TOPIC: Data Structures | ocaml 04_data_structures.ml *)

(* TODO 1: List — fundamental functional data structure *)
(*   [1; 2; 3] |> List.length          (* 3 *)
     [1; 2; 3] |> List.rev             (* [3;2;1] *)
     [1;2] @ [3;4]                      (* append: O(n) *)
     List.nth [10;20;30] 1             (* 20 — O(n) *)
     List.sort compare [3;1;2]         (* [1;2;3] *)
     List.sort_uniq compare [3;1;2;1]  (* [1;2;3] *)  *)

(* TODO 2: Array — fixed-size, O(1) access, mutable *)
(*   let arr = [|1; 2; 3; 4; 5|]
     arr.(2)                            (* 3 — 0-indexed *)
     arr.(2) <- 99                      (* mutate *)
     Array.length arr                  (* 5 *)
     Array.map (fun x -> x * 2) arr
     Array.sort compare arr  *)

(* TODO 3: Hashtbl — mutable hash table *)
(*   let h = Hashtbl.create 16 in       (* initial size hint *)
     Hashtbl.add h "alice" 30;
     Hashtbl.add h "bob" 25;
     Hashtbl.find h "alice";           (* 30; raises Not_found if absent *)
     Hashtbl.find_opt h "carol";       (* None *)
     Hashtbl.mem h "bob";             (* true *)
     Hashtbl.remove h "bob";
     Hashtbl.iter (fun k v -> Printf.printf "%s: %d\n" k v) h  *)

(* TODO 4: Stack from stdlib *)
(*   let s = Stack.create () in
     Stack.push 1 s; Stack.push 2 s; Stack.push 3 s;
     Stack.pop s;                      (* 3 — LIFO *)
     Stack.top s;                      (* 2 — peek without removing *)
     Stack.is_empty s  *)

(* TODO 5: Queue from stdlib *)
(*   let q = Queue.create () in
     Queue.push 1 q; Queue.push 2 q;
     Queue.pop q;                      (* 1 — FIFO *)
     Queue.peek q;                     (* 2 *)  *)

(* TODO 6: Custom linked list type *)
(*   type 'a node = { value: 'a; mutable next: 'a node option }
     type 'a linked_list = { mutable head: 'a node option; mutable length: int }
     let create () = { head = None; length = 0 }
     let push_front lst v =
       let node = { value = v; next = lst.head } in
       lst.head <- Some node;
       lst.length <- lst.length + 1  *)

(* TODO 7: Map module — purely functional balanced BST map *)
(*   module StringMap = Map.Make(String)
     let m = StringMap.empty
           |> StringMap.add "alice" 30
           |> StringMap.add "bob" 25
     StringMap.find "alice" m             (* 30 *)
     StringMap.mem "carol" m             (* false *)
     StringMap.bindings m                (* association list *)  *)

(* TODO 8: Set module — purely functional balanced BST set *)
(*   module IntSet = Set.Make(Int)
     let s = IntSet.of_list [3; 1; 4; 1; 5; 9; 2; 6]
     IntSet.mem 4 s                      (* true *)
     IntSet.min_elt s                    (* 1 *)
     IntSet.inter (IntSet.of_list [1;2;3]) (IntSet.of_list [2;3;4])  *)

let () = print_endline "TODO: implement data structure exercises"
